src/codegen/subexpression_elimination/mod.rs

```diff
@@ -48,10 +51,11 @@ pub struct AvailableExpression {
 
 /// Each BasicExpression is a graph node
 #[derive(Clone)]
-pub struct BasicExpression {
+pub struct BasicExpression<'a> {
     expr_type: ExpressionType,
     expression_id: NodeId,
-    children: HashMap<NodeId, Rc<RefCell<BasicExpression>>>,
+    children: HashMap<NodeId, Rc<RefCell<BasicExpression<'a>>>>,
+    pub reference: &'a Expression,
```

Comment:
I don't know what this new field is for. All the fields could use with (doc) comments